orc8r/cloud/go/services/state/client_api.go

```diff
@@ -103,7 +112,27 @@ func SearchStates(networkID string, typeFilter []string, keyFilter []string, key
 	if err != nil {
 		return nil, err
 	}
-	return state_types.MakeStatesByID(res.States, serdes)
+
+	states, err := state_types.MakeStatesByID(res.States, serdes)
+	if err != nil {
+		return nil, err
+	}
+
+	// check if all state IDs have the exact IMSI prefix (if keyPrefix was specified)
+	if !funk.IsEmpty(keyPrefix) {
+		for stateID := range states {
+			matches := IMSIKeyRe.FindStringSubmatch(stateID.DeviceID)
+			if len(matches) != IMSIExpectedMatchCount {
+				glog.Errorf("state device ID %s did not match IMSI-prefixed regex", stateID.DeviceID)
+				continue
+			}
+			if matches[1] != *keyPrefix {
+				delete(states, stateID)
+			}
+		}
+	}
+
+	return states, nil
```

Comment:
This is the wrong place for the post-filtering. That is, the state service is doing it's job by, in a domain-agnostic way, providing key-prefix search functionality. So it's actually in the subscriberdb handlers (e.g. https://github.com/magma/magma/blob/23d8f7e37e9d66446aaa4f1b16b67f3f4f2b99ea/lte/cloud/go/services/subscriberdb/obsidian/handlers/handlers.go#L408) that we need to handle post-filtering.

We call `state.SearchStates` in a couple places in that handlers file, and we need to (a) test each spot and (b) filter/fix each of those locations where it's currently broken

Feel free to send me a message if you want more info